Issue 256 - move camsl to geo_point table

jo-asplin-met-no opened the pull request:
Fixes [Issue 256](https://github.com/EUMETNET/e-soh/issues/256).

jo-asplin-met-no committed:
Initial attempt at moving camsl to geo_point table, but integration tests fail

jo-asplin-met-no committed:
Removed unused function

jo-asplin-met-no committed:
Fixed bug and typo

lukas-phaf reviewed (commented):
I did a high level review, focusing mostly on the protobuf and postgres migrations. I have added some comments around these, and will try to get to the golang code later.

A general question is how we test this, as there is no ingestion and/or API support to see if inserting and querying actually works (unless I am missing something). 

Ideally, we would have the ingestion and API in the same PR, so that we can add some `camsl`'s to the test dataset, and confirm that everything works by testing that the data shows up in the API, preferable using the `camsl` range query.

lukas-phaf reviewed (commented):
Note to also fix `just load`. Gives concurrency errors...

lukas-phaf commented on datastore/datastore/storagebackend/postgresql/putobservations.go:
```diff
@@ -366,21 +370,21 @@ func getGeoPointIDs(db *sql.DB, observations []*datastore.Metadata1) (map[GeoPoi
 	UNION
-	SELECT c.id, point FROM input_rows
-	JOIN geo_point c USING (point)
+	SELECT c.id, point, camsl FROM input_rows
+	JOIN geo_point c USING (point, camsl)
```
Looks like this fixes the "concurrency" issue on the point table during data load, and the integration test.

jo-asplin-met-no replied:
Interesting, will look into it.

jo-asplin-met-no replied:
Looks good!

jo-asplin-met-no committed:
Merge branch 'main' into issue_256_move_camsl_to_geo_point_table

jo-asplin-met-no committed:
Represented CAMSL as int64 instead of int32

jo-asplin-met-no committed:
Added services to targets test and ingest-test

This change ensures that 'just test' and 'just ingest-test'
can be run without first running 'just up'.

So for example: instead of doing
$ just destroy
$ just up
$ just test

you can now simply do
$ just destroy
$ just test

jo-asplin-met-no committed:
Avoided modifying field numbers of existing fields

jo-asplin-met-no committed:
Represented CAMSL as int32 instead of int64 (again!)

jo-asplin-met-no committed:
Added better/safer migrations for moving camsl to geo_point table

jo-asplin-met-no commented on datastore/migrate/data/migrations/1747651104_camsl.up.sql:
```diff
@@ -1 +1,9 @@
+ALTER TABLE geo_point ADD COLUMN camsl INTEGER;
+
+-- drop UNIQUE constraint of 'point' column
+-- WARNING: we assume that the constraint name is the correct one (it was never explicitly set)
```
Good question. Need to look into it again.

jo-asplin-met-no committed:
Fixed example

jo-asplin-met-no commented on datastore/datastore/storagebackend/postgresql/putobservations.examples.sql:
Fixed! (I think)

jo-asplin-met-no committed:
Merge branch 'main' into issue_256_move_camsl_to_geo_point_table

jo-asplin-met-no commented:
I got stuck on `just ingest-load` failing due to a concurrency error. I think @lukas-phaf needs to take over this particular problem since he's the one who implemented the ingest optimization in [putobservations.go](https://github.com/EUMETNET/e-soh/blob/issue_256_move_camsl_to_geo_point_table/datastore/datastore/storagebackend/postgresql/putobservations.go). For reasons that are beyond me, the seemingly harmless addition of `camsl` next to `point` in the `geo_point` table now causes the complex INSERT statement in the function `getGeoPointIDs` to consistently fail in all 3 attempts, thus resulting in the error.

lukas-phaf commented:
> I got stuck on `just ingest-load` failing due to a concurrency error. I think @lukas-phaf needs to take over this particular problem since he's the one who implemented the ingest optimization in [putobservations.go](https://github.com/EUMETNET/e-soh/blob/issue_256_move_camsl_to_geo_point_table/datastore/datastore/storagebackend/postgresql/putobservations.go). For reasons that are beyond me, the seemingly harmless addition of `camsl` next to `point` in the `geo_point` table now causes the complex INSERT statement in the function `getGeoPointIDs` to consistently fail in all 3 attempts, thus resulting in the error.

Fair enough. The problem is not obvious to me by just reading the statement. I will try to look at this tomorrow.

lukas-phaf committed:
Fix JOIN so it can deal with NULL.

lukas-phaf commented:
> I got stuck on `just ingest-load` failing due to a concurrency error. I think @lukas-phaf needs to take over this particular problem since he's the one who implemented the ingest optimization in [putobservations.go](https://github.com/EUMETNET/e-soh/blob/issue_256_move_camsl_to_geo_point_table/datastore/datastore/storagebackend/postgresql/putobservations.go). For reasons that are beyond me, the seemingly harmless addition of `camsl` next to `point` in the `geo_point` table now causes the complex INSERT statement in the function `getGeoPointIDs` to consistently fail in all 3 attempts, thus resulting in the error.

I committed a fix, which make the `just load integration` work for me.
The issue was that the JOIN at the end of the query does not deal correctly with NULL's (in `camsl`). I replaced the implicit join condition with an explicit one that deals properly with NULL's.

I did not yet update the file with example queries.

jo-asplin-met-no committed:
Merge branch 'issue_256_move_camsl_to_geo_point_table' of github.com:EUMETNET/e-soh into issue_256_move_camsl_to_geo_point_table

jo-asplin-met-no committed:
Avoided referencing removed column

jo-asplin-met-no committed:
Merge branch 'main' into issue_256_move_camsl_to_geo_point_table
